src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1173 @@
+            None,
+            Some(pin_auth.clone()),
+        );
+        assert_eq!(response, Ok(()));
+        assert_eq!(persistent_store.min_pin_length(), min_pin_length);
+        let response = pin_protocol_v1.process_set_min_pin_length(
+            &mut persistent_store,
+            7,
+            None,
+            Some(pin_auth),
+        );
+        assert_eq!(
+            response,
+            Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION)
+        );
+        assert_eq!(persistent_store.min_pin_length(), min_pin_length);
+    }
+
+    #[test]
+    fn test_process() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+        let mut pin_protocol_v1 = PinProtocolV1::new(&mut rng);
+        let client_pin_params = AuthenticatorClientPinParameters {
+            pin_protocol: 1,
+            sub_command: ClientPinSubCommand::GetPinRetries,
+            key_agreement: None,
+            pin_auth: None,
+            new_pin_enc: None,
+            pin_hash_enc: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length_rp_ids: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        };
+        assert!(pin_protocol_v1
+            .process_subcommand(&mut rng, &mut persistent_store, client_pin_params)
+            .is_ok());
+
+        let client_pin_params = AuthenticatorClientPinParameters {
+            pin_protocol: 2,
+            sub_command: ClientPinSubCommand::GetPinRetries,
+            key_agreement: None,
+            pin_auth: None,
+            new_pin_enc: None,
+            pin_hash_enc: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length_rp_ids: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        };
+        #[cfg(not(feature = "with_ctap2_1"))]
+        let error_code = Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID;
+        #[cfg(feature = "with_ctap2_1")]
+        let error_code = Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER;
+        assert_eq!(
+            pin_protocol_v1.process_subcommand(&mut rng, &mut persistent_store, client_pin_params),
+            Err(error_code)
+        );
+    }
+
+    #[test]
+    fn test_check_and_store_new_pin() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+        let shared_secret = [0x88; 32];
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+
+        // The PIN "1234" should be accepted.
```

Comment:
For these tests, it's not obvious that the pin is "1234" or "123" from the encrypted golden value. So I'd suggest presenting the tests more in terms of the decrypted blob (e.g. `[1, 2, 3, 4, 0, 0, 0, ...]`), which is then encrypted with `aes_enc_key` available in this function, and then passed to `check_and_store_new_pin`.

Also, `check_and_store_new_pin` returning only a boolean doesn't help to check that the rejection reason is indeed what is mentioned in each comment. It would be nice to check an error code and/or check what the storage contains afterwards. In particular have a more complete scenario (maybe a separate unit test) where we try to use the new pin, and confirm that the old pin doesn't work (or the other way around if the new pin was rejected as invalid).

I think it's quite important to have clear tests for this critical scenario (authentication).

That's also why I think having separate functions for the "encrypt pin" and for the "decrypt pin" logic would make things much clearer for testing.

Reply:
I introduced `decrypt_pin` and a separate test. The tests for `test_check_and_store_new_pin` now also look into the storage. `encrypt_pin` is only extended for creating more flexible `new_pin_enc`.